pkg/ivd/ivd_protected_entity_type_manager_util.go

```diff
@@ -0,0 +1,204 @@
+				VSphereUser: user,
+			},
+			EntityMetadata: metadataList,
+		},
+		BackingObjectDetails: &cnstypes.CnsBlockBackingDetails{
+			CnsBackingObjectDetails: cnstypes.CnsBackingObjectDetails{
+				CapacityInMb: md.VirtualStorageObject.Config.CapacityInMB,
+			},
+		},
+	}
+
+	cnsVolumeCreateSpecList = append(cnsVolumeCreateSpecList, cnsVolumeCreateSpec)
+	logger.Debugf("Provisioning volume using the spec: %v", cnsVolumeCreateSpec)
+
+	// provision volume using CNS API
+	cnsClient, err := cns.NewClient(ctx, client.Client)
+	if err != nil {
+		return "", err
+	}
+	createTask, err := cnsClient.CreateVolume(ctx, cnsVolumeCreateSpecList)
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to create volume. Error: %+v", err)
+		return "", err
+	}
+	createTaskInfo, err := cns.GetTaskInfo(ctx, createTask)
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to create volume. Error: %+v", err)
+		return "", err
+	}
+	createTaskResult, err := cns.GetTaskResult(ctx, createTaskInfo)
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to create volume. Error: %+v", err)
+		return "", err
+	}
+	if createTaskResult == nil {
+		err := errors.New("Empty create task results")
+		logger.Error(err.Error())
+		return "", err
+	}
+	createVolumeOperationRes := createTaskResult.GetCnsVolumeOperationResult()
+	if createVolumeOperationRes.Fault != nil {
+		logger.Errorf("Failed to create volume: fault=%+v", createVolumeOperationRes.Fault)
+		return "", errors.New(createVolumeOperationRes.Fault.LocalizedMessage)
+	}
+
+	volumeId := createVolumeOperationRes.VolumeId.Id
+	logger.Infof("CNS volume, %v, created", volumeId)
+	return volumeId, nil
+}
+
+func filterLabelsFromMetadata(md Metadata, prefix string, logger logrus.FieldLogger) Metadata {
+	// prefix: cns.containerCluster
+	var kvsList []vim25types.KeyValue
+
+	logger.Debugf("labels of CNS voluem before filtering ones with certain prefix, %v: %v", prefix, md.ExtendedMetadata)
+
+	for _, label := range md.ExtendedMetadata {
+		if !strings.HasPrefix(label.Key, prefix) {
+			kvsList = append(kvsList, vim25types.KeyValue {
+				Key: label.Key,
+				Value: label.Value,
+			})
+		}
+	}
+	md.ExtendedMetadata = kvsList
+
+	logger.Debugf("labels of CNS voluem after filtering ones with certain prefix, %v: %v", prefix, md.ExtendedMetadata)
+
+	return md
+}
+
+func CreateCnsVolumeInCluster(ctx context.Context, client *govmomi.Client, md Metadata, logger logrus.FieldLogger) (string, error) {
+	logger.Infof("CreateCnsVolumeInCluster called with args, Metadata: %v", md)
+
+	// Retrieving cluster id and vSphere user
+	logger.Debug("Retrieving cluster id and vSphere user required by provisioning volume")
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to get k8s inClusterConfig")
+		return "", err
+	}
+
+	// To enable cross-cluster restore, need to filter out the cns specific labels, i.e. prefix: cns, in md
+	md = filterLabelsFromMetadata(md, "cns", logger)
```

Comment:
This is good, would you add this logic to the local clone and create from snapshot paths in copyInt?

Reply:
sure.